Editor: Optimize story reducers using immer library

barklund opened the pull request:
## Context

This converts the story reducers to using immer to make them easier to read and write (purely DX).

This uses all the new and extended unit tests for the reducers to make sure the conversion is done correctly.

## User-facing changes
 _None_.

## Testing Instructions

<!-- ignore-task-list-start -->
- [x] This is a non-user-facing change and requires no QA
<!-- ignore-task-list-end -->

## Checklist

<!-- Check these after PR creation -->

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #11723

Commit:
Added immer as a dependency and converted some reducers

Commit:
Merge branch 'main' into try/immer

# Conflicts:
#	packages/story-editor/src/app/story/useStoryReducer/reducers/addElements.js
#	packages/story-editor/src/app/story/useStoryReducer/reducers/combineElements.js
#	packages/story-editor/src/app/story/useStoryReducer/reducers/copySelectedElement.js
#	packages/story-editor/src/app/story/useStoryReducer/reducers/deleteElementsByResourceId.js
#	packages/story-editor/src/app/story/useStoryReducer/reducers/setBackgroundElement.js
#	packages/story-editor/src/app/story/useStoryReducer/test/deleteElementsByResourceId.js

Commit:
Fixed a test after bad merge and fixed an actual error in one immerplementation

Commit:
Rewritten remaining reducers to immer

Commit:
Added a comment

Commit:
Simplifying reducer a lot

barklund commented on packages/story-editor/src/app/story/useStoryReducer/reducers/utils.js:
```diff
@@ -87,9 +87,13 @@ export function updateElementWithUpdater(element, properties) {
+  }
+  Object.assign(element, allowedProperties);
+  return null;
 }
```
Some of the remaining functions in this file could also be updated so that they work on the passed objects rather than return immutable clones.

Commit:
Merge branch 'main' into try/11723-immer-reducers

# Conflicts:
#	packages/story-editor/src/app/story/useStoryReducer/reducers/addElements.js
#	packages/story-editor/src/app/story/useStoryReducer/reducers/arrangeElement.js
#	packages/story-editor/src/app/story/useStoryReducer/reducers/setSelectedElements.js

Commit:
Updated reducer and test for arrange element with groups

Commit:
Added tests for new group reducers (and fixed some minor things)

Note that I think there is a bug, but I will see ask around if there actually is one?

merapi commented on packages/story-editor/src/app/story/useStoryReducer/reducers/arrangeElement.js:
```diff
@@ -86,38 +89,21 @@ function arrangeElement(state, { elementId, position, groupId = false }) {
+  if (groupId) {
+    page.elements[currentPosition].groupId = groupId;
+  } else {
+    delete page.elements[currentPosition].groupId;
```
Right now the `groupId` can be either a `string` = actual groupId (move into the group), `null/undefined` = remove the groupId (move out of the group) or `false` = don't update the id (move within the group/non-group).

` * @param {string|boolean} payload.groupId New group id (when moving in the Layers Panel), false to skip updating.`

merapi committed:
Fix missing groupId

merapi committed:
The test should now pass

merapi committed:
Better fix?

Commit:
Added proper tests and docs for group id on arrange element

timarney commented on packages/story-editor/src/app/story/useStoryReducer/reducers/combineElements.js:
```diff
@@ -53,115 +54,87 @@ import { removeAnimationsWithElementIds } from './utils';
+      // Update reference to second element
+      .map((el) => (el.id === secondId ? newElement : el));
+
+    // First element should always be the image getting applied to
```
These comments are really helpful - to give some back story into what is happening.

💯

Commit:
Updated new group functions to immer reducers

Also found a better structure for reducers giving access to inner "mutating" function for nested producers

Commit:
Updated all reducers to new and better structure

Commit:
Updated reducer after simplifying it

merapi reviewed (approved):
Looks good and should be safe to merge thanks to the good test coverage 💯 👍🏻

Commit:
Merge branch 'main' into try/11723-immer-reducers

Commit:
Added tests for selection after combine and fixed bug

Commit:
A fix to the way styles are pasted

This is required because updateElements now calls its callback async, and not sync as before, so we need to store the props locally before removing them.

Commit:
Fixed test that was actually broken and relied on broken behavior
